agent/php_user_instrument_hashmap.c

```diff
@@ -0,0 +1,315 @@
+  /* for efficiency wraprec is used as both: key and a value */
+  nruserfn_t *wraprec;
+} nr_wraprecs_bucket_t;
+
+typedef struct _nr_php_wraprec_hashmap {
+  nr_php_wraprec_hashmap_dtor_fn_t dtor_func;
+  size_t log2_num_buckets;
+  nr_wraprecs_bucket_t** buckets;
+  size_t elements;
+} nr_php_wraprec_hashmap_t;
+
+static inline size_t nr_count_buckets(const nr_php_wraprec_hashmap_t* hashmap) {
+  return (size_t)(1 << hashmap->log2_num_buckets);
+}
+
+static nr_php_wraprec_hashmap_t* nr_php_wraprec_hashmap_create_internal(size_t log2_num_buckets,
+                                         nr_php_wraprec_hashmap_dtor_fn_t dtor_fn) {
+  nr_php_wraprec_hashmap_t* hashmap;
+
+  if (0 == log2_num_buckets) {
+    /*
+     * Encode the default value in one place: namely, here.
+     */
+    log2_num_buckets = 8;
+  } else if (log2_num_buckets > 24) {
+    /*
+     * Basic sanity check: it's extremely unlikely that we'll ever need a
+     * hashmap for the user function wraprecs that has more than 2^24 buckets.
+     */
+    log2_num_buckets = 24;
+  }
+
+  hashmap = (nr_php_wraprec_hashmap_t*)nr_malloc(sizeof(nr_php_wraprec_hashmap_t));
+  hashmap->dtor_func = dtor_fn;
+  hashmap->log2_num_buckets = log2_num_buckets;
+  hashmap->buckets = (nr_wraprecs_bucket_t**)nr_calloc(
+      (1 << log2_num_buckets), sizeof(nr_wraprecs_bucket_t*));
+  hashmap->elements = 0;
+
+  return hashmap;
+}
+
+nr_php_wraprec_hashmap_t* nr_php_wraprec_hashmap_create_buckets(size_t buckets, nr_php_wraprec_hashmap_dtor_fn_t dtor_fn) {
+
+  size_t actual_buckets;
+
+  if (0 == buckets) {
+    actual_buckets = 0;
+  } else {
+    /*
+     * Calculate ceil(log2(buckets)), since that's what
+     * nr_hashmap_create_internal requires.
+     */
+
+    actual_buckets = 1;
+    while (((size_t)(1 << actual_buckets)) < buckets) {
+      actual_buckets += 1;
+    }
+  }
+
+  return nr_php_wraprec_hashmap_create_internal(actual_buckets, dtor_fn);
+}
+
+static void nr_destroy_wraprecs_bucket(nr_wraprecs_bucket_t** bucket_ptr,
+                               nr_php_wraprec_hashmap_dtor_fn_t dtor_func) {
+  nr_wraprecs_bucket_t* bucket = *bucket_ptr;
+
+  if (dtor_func) {
+    (dtor_func)(bucket->wraprec);
+  }
+  nr_realfree((void**)bucket_ptr);
+}
+
+nr_php_wraprec_hashmap_stats_t nr_php_wraprec_hashmap_destroy(nr_php_wraprec_hashmap_t** hashmap_ptr) {
+  nr_php_wraprec_hashmap_stats_t stats = {};
+  size_t count;
+  nr_php_wraprec_hashmap_t* hashmap;
+  size_t i;
+  size_t collisions;
+
+  if ((NULL == hashmap_ptr) || (NULL == *hashmap_ptr)) {
+    return stats;
+  }
+  hashmap = *hashmap_ptr;
+
+  stats.elements = hashmap->elements;
+  stats.collisions_min = stats.elements;
+
+  count = nr_count_buckets(hashmap);
+  for (i = 0; i < count; i++) {
+    nr_wraprecs_bucket_t* bucket = hashmap->buckets[i];
+    collisions = 0;
+
+    if (bucket) {
+      stats.buckets_used++;
+    }
+
+    while (bucket) {
+      nr_wraprecs_bucket_t* next = bucket->next;
+
+      collisions++;
```

Comment:
Is it intended to count a collision even if the bucket list is of length one?

Reply:
Yes it is intentional. [This condition](https://github.com/misiekn/newrelic-php-agent/blob/refactor/wraprec-lookup/agent/php_user_instrument_hashmap.c#L126) checks if the `collisions` counter is greater than 1 and only then updates collisions mean stats. I guess maybe the variable name is not the most fortunate. Do you think `bucket_items_cnt` is better variable name?

Reply:
Yeah, I think that would be more clear when reading this code.

Reply:
@ZNeumann  Please see a09d04db3ce424dcb6012891c85f331c7a31d76a. If this is acceptable, please mark this conversation as resolved.